[LoopUnroll] Fix division by zero

jdenny-ornl opened the pull request:
PR #159163's probability computation for epilogue loops does not handle the possibility of an original loop probability of one. Runtime loop unrolling does not make sense for such an infinite loop, and a division by zero results.  This patch works around that case.

Issue #165998.

jdenny-ornl committed:
[LoopUnroll] Fix division by zero

PR #159163's probability computation for epilogue loops does not
handle the possibility of an original loop probability of one.
Runtime loop unrolling does not make sense for such an infinite loop,
and a division by zero results.  This patch works around that case.

Issue #165998.

MatzeB reviewed (approved):
Code change LGTM.

Not sure it's worth speculating about the situation in the comment to a random function... I think I remember the sample profile loader avoiding 1:0 probabilities even if we do have not picked up any samples on one side (but it's statistics, and it can happen when there is few/not enough samples for the loop at hand); but given the IR verifier isn't rejecting this case and who knows what other sources or passes produce probability annotationes, seems fair/necessary to handle this case.

MatzeB commented on llvm/lib/Transforms/Utils/LoopUnrollRuntime.cpp:
```diff
@@ -200,8 +200,17 @@ static void ConnectProlog(Loop *L, Value *BECount, unsigned Count,
 probOfNextInRemainder(BranchProbability OriginalLoopProb, unsigned N) {
+  if (OriginalLoopProb == BranchProbability::getOne())
+    return BranchProbability::getOne();
+
```
Not a strong opinion here, but for my taste I'd rather have the talk about special cases within the function (rather than in the "API" documentation in front of the function) and even then rather discuss in reviews, discourse etc. what the values could mean rather than particularily here.

(But I realize this is a bit of a subjective thing; so do whatever you think is right)

jdenny-ornl commented:
> Code change LGTM.

Thanks.

> Not sure it's worth speculating about the situation in the comment to a random function...

I'm not sure I understand what you mean.  I wrote the probOfNextInRemainder function to compute correct probabilities for epilogue remainder loops.  I think it is worthwhile to point out that there is *no* reasonable probability for that case when the original loop is infinite, so the probability we return is an arbitrary choice.

Andarwinux commented:
Thanks! I can confirm the problem has been fixed.

MatzeB commented:
¯\_(ツ)_/¯ just felt like a lot of text (which to me felt like it may as easily confuse readers as it may help). Is the profile really "bad" in that case? I think having a 1:0 probability is still possible either because it's actually infinite, or because the difference is too small to represent. But as long as the IR verifier accepts it, it's not "bad" in my book... But as I said arguing about source code comments is highly subjective, so do with my comment as you wish or ignore...

jdenny-ornl committed:
Try to improve code comments

jdenny-ornl commented:
> ¯_(ツ)_/¯ just felt like a lot of text (which to me felt like it may as easily confuse readers as it may help). Is the profile really "bad" in that case? I think having a 1:0 probability is still possible either because it's actually infinite, or because the difference is too small to represent.

You convinced me.  I moved it into the function body.  I reworded it to focus more on our attempt to derive a meaningful result from the probability rather than on whether it was truly wrong.

> But as I said arguing about source code comments is highly subjective, so do with my comment as you wish or ignore...

Thanks for raising it.

jdenny-ornl committed:
Be even more careful with comments

And trigger pre-commit CI again because it was failing on a test that
surely this PR did not affect.
